### ISSUE TITLE
docs: qualitative figure set and measured benchmark harness for the showcase

### ISSUE THREAD

**Issue by KhoiDOO** [KhoiDOO/conquer3d#26]
The showcase needs qualitative figures that demonstrate what the operators
actually produce, and a benchmark harness that measures the library rather than
quoting numbers from memory.

**Scope**

- Figures for the sparse-grid construction pipeline, Morton ordering, and BVH
  ray queries (triangle hits and voxel traversal).
- A resolution ladder to $2048^3$ with Chamfer/Hausdorff error at every step.
- A reproducible benchmark script covering extraction, sign modes, distance
  operators and pipeline setup.
- Site layout fixes surfaced while reviewing the figures.

**Comment by KhoiDOO** [comment]
Implemented in #25 (branch `docs/showcase-refresh`, 9 commits).

## 1. API usage

**Regenerating everything.** Both pipelines are standalone and take a figure
name to rebuild just one:

```bash
python docs/_figures/make_figures.py            # all figures
python docs/_figures/make_figures.py meshbvh    # one figure
python docs/_figures/make_benchmarks.py         # plots + benchmarks.json
python docs/generate.py --strict                # site; fails on any regression
```

**The ray-query figure**, which is the clearest small example of the query API.
`MeshBVH.get_ray_intersection` returns a 4-tuple — note that element `[1]` is
the *triangle* index, not a hit flag:

```python
import torch
from conquer3d._C import BVH, MeshBVH
from conquer3d.creation import create_sphere
from conquer3d.data_structure import TriangleMesh, create_voxel_grid_from_tmesh

# create_sphere(sectors, stacks, radius) -> (verts, tris); note the arg order
verts, tris = create_sphere(26, 16, 1.0)
verts = verts.cuda().float().contiguous()
tris = tris.cuda().int().contiguous()

origin = torch.tensor([[-3.0, 0.84, 0.0]], device="cuda")   # (R, 3)
direction = torch.tensor([[1.0, 0.0, 0.0]], device="cuda")  # (R, 3), need not be unit

tv = verts[tris.long()]                                     # (T, 3, 3)
mesh_bvh = MeshBVH(tv.min(1).values.contiguous(), tv.max(1).values.contiguous())

query_ids, tri_ids, hit_points, distances = mesh_bvh.get_ray_intersection(
    origin, direction, verts, tris, True,                   # True -> return distances
)
# tri_ids: int64 (H,) indices into `tris`; hit_points: float32 (H, 3)
```

Traversing the *cells* the same ray crosses is the same call shape against a
`BVH` built over the voxel AABBs:

```python
tmesh = TriangleMesh(verts, tris)
gv, vox = create_voxel_grid_from_tmesh(
    grid_min=[-1.3] * 3, grid_max=[1.3] * 3, res=[22] * 3,
    tmesh=tmesh, pad=1, return_normals=False,
)[:2]

corners = gv[vox.long()]                                    # (C, 8, 3)
voxel_bvh = BVH(corners.min(1).values.contiguous(), corners.max(1).values.contiguous())
_, cell_ids = voxel_bvh.query_ray(origin, direction)        # max_capacity defaults to 1e7
```

**Surface error at a given resolution**, used for the resolution ladder:

```python
from conquer3d.ops import chamfer_distance, hausdorff_distance

ref = tmesh.sample_points(200_000)          # area-weighted
got = extracted.sample_points(200_000)
cd = chamfer_distance(got, ref)             # symmetric by default
hd = hausdorff_distance(got, ref)
```

## 2. Advantages

- **Nothing in the figures is staged.** Every panel is real operator output on
  the bundled benchmark assets, regenerated from source, so a figure cannot
  quietly drift from library behaviour.
- **Sparse throughout.** Every extraction figure runs on a narrow-band grid from
  `create_voxel_grid_from_tmesh`; at $1024^3$ that is 5,148,906 cells, **0.48%**
  of the 1.07B a dense volume would hold. The figures now state this explicitly,
  since a picture of a mesh gives no hint that no dense volume was ever held.
- **Benchmarks are honest.** CUDA events around the operator alone, median of 7
  after 2 warm-ups, on a single machine with the environment recorded in
  `benchmarks.json` (RTX 4090, torch 2.8.0+cu128, CUDA 12.8, Python 3.10.14).
- **Coverage is enforced, not asserted.** `--strict` fails the build on an
  undocumented symbol, a dead internal link, or a missing native tier.

## 3. Disadvantages & trade-offs

- **The error ladder hits a measurement floor, not a quality floor.** Chamfer
  falls 8.54e-3 → 4.81e-3 from $64^3$ to $1024^3$, then flattens: 4.797e-3 at
  $1536^3$ and essentially the same at $2048^3$. That plateau is the 200k-sample
  Monte Carlo floor of the metric itself, not the surface ceasing to improve.
  Reading it as convergence would be wrong.
- **`build_flood_fill_data` cannot reach high resolutions.** It allocates a dense
  `int32` volume — $1536^3$ is 13.5 GiB and OOMs on a 24 GB card. The figures
  switch to `build_flood_fill_cf_data` with `sign_mode=5` at $\geq 1024^3$.
- **`MAX_K = 32`** bounds KD-tree k-NN: the per-thread priority queue lives in
  registers, so k > 32 raises rather than spilling to local memory.
- **Setup dominates extraction by two orders of magnitude.** At $1024^3$ the
  flood fill costs 799.74 ms against 2.62 ms for DMC. Benchmarking extraction
  alone is honest only because the setup is amortised across extractions of the
  same field.
- Figure assets add ~2.4 MB of WebP to the repository. Line art and text needed
  *lossless* WebP — lossy smeared matplotlib glyphs into blocks.

## 4. Empirical benchmarks

RTX 4090, torch 2.8.0+cu128, CUDA 12.8. Fandisk at $1024^3$, 5,148,906 active
cells. CUDA events, median of 7 runs after 2 warm-ups.

| Algorithm | Output | Vertices | Faces | Latency | Throughput |
| :--- | :--- | ---: | ---: | ---: | ---: |
| Dual Marching Cubes | Triangles | 1,716,384 | 3,432,764 | **2.62 ms** | **1,310M f/s** |
| DMC (pure quads) | Quads | 1,716,384 | 1,716,382 | **2.51 ms** | 684M q/s |
| MC Asymptotic | Triangles | 1,716,382 | 3,432,760 | 2.71 ms | 1,267M f/s |
| Dual Contouring | Triangles | 1,716,384 | 3,432,764 | 3.88 ms | 885M f/s |
| Marching Cubes | Triangles | 1,716,382 | 3,432,760 | 7.88 ms | 436M f/s |
| Marching Tetrahedra | Triangles | 6,113,918 | 12,227,832 | 44.56 ms | 274M f/s |

**This contradicts the previously published table.** MCA was listed at 16.42 ms
and measures **2.71 ms** — a 6× discrepancy that should be confirmed
independently before the next release. DC measured 3.88 ms against a published
4.79 ms; DMC 2.62 ms against 3.44 ms.

Pipeline setup, Fandisk (`sparsity` = share of the dense grid actually allocated):

| Res | Grid build | Flood fill | SDF query | Active cells | Sparsity |
| ---: | ---: | ---: | ---: | ---: | ---: |
| $128^3$ | 4.16 ms | 3.60 ms | 0.29 ms | 79,472 | 3.79% |
| $256^3$ | 5.23 ms | 9.10 ms | 0.31 ms | 319,357 | 1.90% |
| $512^3$ | 14.96 ms | 34.39 ms | 0.63 ms | 1,280,303 | 0.95% |
| $1024^3$ | 66.64 ms | 799.74 ms | 1.83 ms | 5,148,906 | 0.48% |

Sign modes at 13M query points — the six modes land within **10%** of each
other (999–1096 ms, 11.87–13.01M pts/s), with flood fill and coarse-fine fill
fastest and ray parity slowest. Sweep run at 1/2/3/5/7/11/13/17/19/23M points.

Distance operators at 1M points: one-sided Chamfer 11.64 ms (85.9M pts/s),
symmetric 23.21 ms (43.1M pts/s); Hausdorff is within noise of Chamfer at
11.63 / 23.23 ms, which is expected since both share the same k-NN traversal
and differ only in the reduction.

Surface error vs resolution (Armadillo, 200k samples):

| Res | Faces | Chamfer | Hausdorff |
| ---: | ---: | ---: | ---: |
| $64^3$ | 12,076 | 8.542e-3 | 4.617e-2 |
| $128^3$ | 49,904 | 5.746e-3 | 1.640e-2 |
| $256^3$ | 204,548 | 4.986e-3 | 1.299e-2 |
| $512^3$ | 825,960 | 4.829e-3 | 1.306e-2 |
| $1024^3$ | 3,317,336 | 4.806e-3 | 1.284e-2 |
| $1536^3$ | 7,472,800 | 4.797e-3 | 1.282e-2 |

## 5. Technical insights

- **A diametric ray can never show both of its hit triangles.** Entry and exit
  face opposite directions, so one is always back-facing. The figure uses a
  near-tangent ray ($y = 0.84$ on a unit sphere) so both hits land on the same
  visible cap. The renderer is Y-up — offsetting on $z$ puts the ray on the
  silhouette instead.
- **Per-face colour requires exploding the mesh.** Highlighting individual
  triangles means per-face vertices (`verts[tris]` reshaped), since vertex
  colours are shared across incident faces.
- **DC and DMC produced bright slivers** until shading was made two-sided: the
  quad-split emits mixed winding, so half of each split shaded as a back face.
  `nrm * where((nrm · view) < 0, -1, 1)` fixes it without touching the topology.
- **`get_quality()` returns scalar (min, mean)**, not a per-face field, so the
  curvature figure uses principal curvature $\kappa_1$ from the Laplace–Beltrami
  operator instead.
- **The site was serving stale CSS.** No cache headers and unversioned asset
  URLs meant a browser kept the stylesheet it already had, so verified fixes did
  not reach the page. Assets now carry a content hash and the `--strict` link
  checker strips the query string before resolving.
- **Two sign-mode figure "failures" were preview artifacts.** Text that appeared
  as grey blocks was downscaling plus a transparent background; both verified
  correct by cropping at native resolution and compositing on white and dark.

**Comment by KhoiDOO** [comment]
Implemented and covered by PR #25.
